torchdata/datapipes/iter/util/tfrecordloader.py

```diff
@@ -0,0 +1,241 @@
+    return result
+
+
+def parse_tfrecord_sequence_example(example, spec: Optional[ExampleSpec]) -> Example:
+    # Parse context features
+    result = cast(Example, _parse_tfrecord_features(example.context, spec))
+
+    # Parse feature lists
+    feature_lists_keys = None if spec is None else set(spec.keys()) - set(result.keys())
+    features = example.feature_lists.feature_list
+    for key in features.keys():
+        if feature_lists_keys is not None and key not in feature_lists_keys:
+            continue
+        feature_spec = None if spec is None else spec[key]
+        feature = features[key].feature
+        if key in result:
+            raise RuntimeError(
+                "TFRecord example's key {key} is contained in both the context and feature lists. This is not supported."
+            )
+
+        value: Union[torch.Tensor, List[Any]] = list(map(partial(process_feature), feature))
+
+        # For known torch dtypes, we stack the list features
+        if feature_spec is not None and isinstance(feature_spec[1], torch.dtype):
+            value = torch.stack(cast(List[torch.Tensor], value), 0)
+        value = _apply_feature_spec(value, feature_spec)
+        result[key] = value
+    if spec is not None and len(result.keys()) != len(spec.keys()):
+        raise RuntimeError(f"Example is missing some required keys: {sorted(result.keys())} != {sorted(spec.keys())}")
+    return result
+
+
+@functional_datapipe("load_from_tfrecord")
+class TFRecordLoaderIterDataPipe(IterDataPipe[Example]):
+    r"""
+    Opens/decompresses tfrecord binary streams from an Iterable DataPipe which contains tuples of path name and
+    tfrecord binary stream, and yields the stored records (functional name: ``load_from_tfrecord``).
+
+    Args:
+        datapipe: Iterable DataPipe that provides tuples of path name and tfrecord binary stream
+        length: a nominal length of the DataPipe
+
+    Note:
+        The opened file handles will be closed automatically if the default ``DecoderDataPipe``
+        is attached. Otherwise, user should be responsible to close file handles explicitly
+        or let Python's GC close them periodically.
+
+    Example:
+        >>> from torchdata.datapipes.iter import FileLister, FileOpener
+        >>> datapipe1 = FileLister(".", "*.tfrecord")
+        >>> datapipe2 = FileOpener(datapipe1, mode="b")
+        >>> tfrecord_loader_dp = datapipe2.load_from_tfrecord()
+        >>> for example in tfrecord_loader_dp:
+        >>>     print(example)
+    """
+
+    def __init__(
+        self, datapipe: Iterable[Tuple[str, BufferedIOBase]], spec: ExampleSpec = None, length: int = -1
+    ) -> None:
+        super().__init__()
+        _assert_protobuf()
+
+        self.datapipe: Iterable[Tuple[str, BufferedIOBase]] = datapipe
+        self.length: int = length
+        self.spec = spec
+
+    def __iter__(self) -> Iterator[Example]:
+        from . import _tfrecord_example_pb2 as example_pb2
```

Comment:
I believe various TFRecords might have different protibufs, for this you probably need to pass protobuf_schema as parameter of pipe, something like:

```
from . import _tfrecord_example_pb2 as example_pb2
tfrecord_loader_dp = datapipe2.load_from_tfrecord(example_pb2)
```

Reply:
The protobuf schema for Example was kept the same in TF so far. If they changed it, TF would not load older tfrecord files. Therefore, I believe there is no need to pass the protobuf schema explicitly.

Reply:
Since the schema has been consistent I don't think we will need it as a parameter, can you add a comment there stating that is the assumption.

Reply:
I added the comment.

Reply:
nit (mostly for discussion): I understand that TF consistently has the same `protobuf` schema, but we can potentially make this a generic `ProtobufLoaderDataPipe` by accepting schema as an argument (but let the TF one be the default).

Will this be useful for users? What other `protobuf` schema will they use? Or should we just stick with TFRecord only for now? @VitalyFedyunin

Reply:
I would stick with TFRecord only. In the TFRecordLoader we assume the data are Examples in several places. This allows us to load the data in a more "user-friendly" way (e.g. we load FloatList data as tensors and allow the user to reshape them). Therefore, I think it would be better to keep ProtobufLoader separately if there was a need for it.

Reply:
I agree that supporting TFRecord for now but our team worries that we may run into trademark issues if we choose to name it something related to TF. Can you please rename it to `ProtobufRecordLoader` but mention in the description that it, by default, supports loading of TF records?

Reply:
This is a problem. "tfrecord" file describes a sequence of binary strings with headers and protobuf-searialized binary payload. The way the protobuf-serialized records are concatenated in the final tfrecord file is not a part of the protobuf specification. A simple protobuf deserialization could not load a tfrecord file (and would not be efficient, since the whole file would be loaded at once and not sequentially).

Reply:
The "tensorboard" name was used in the PyTorch repo. Could we do something similar in this case? Is the "tfrecord" specification legally protected?

Reply:
It seems like we may want to keep the `TFRecord` name, @VitalyFedyunin what do you think?

Reply:
We are checking with our legal team about this. Will update here when I have more info.

Reply:
We are in the clear with naming and etc. No need to change any names here.